components/Toast/use-toast.tsx

```diff
@@ -0,0 +1,194 @@
+"use client"
+
+// Inspired by react-hot-toast library
+import * as React from "react"
+
+import type {
+	ToastActionElement,
+	ToastProps,
+} from "./toast"
+
+const TOAST_LIMIT = 6
+const TOAST_REMOVE_DELAY = 1000000
+
+type ToasterToast = ToastProps & {
+	id: string
+	title?: React.ReactNode
+	description?: React.ReactNode
+	action?: ToastActionElement
+}
+
+const actionTypes = {
+	ADD_TOAST: "ADD_TOAST",
+	UPDATE_TOAST: "UPDATE_TOAST",
+	DISMISS_TOAST: "DISMISS_TOAST",
+	REMOVE_TOAST: "REMOVE_TOAST",
+} as const
+
+let count = 0
+
+function genId() {
+	count = (count + 1) % Number.MAX_SAFE_INTEGER
+	return count.toString()
+}
+
+type ActionType = typeof actionTypes
+
+type Action =
+	| {
+		type: ActionType["ADD_TOAST"]
+		toast: ToasterToast
+	}
+	| {
+		type: ActionType["UPDATE_TOAST"]
+		toast: Partial<ToasterToast>
+	}
+	| {
+		type: ActionType["DISMISS_TOAST"]
+		toastId?: ToasterToast["id"]
+	}
+	| {
+		type: ActionType["REMOVE_TOAST"]
+		toastId?: ToasterToast["id"]
+	}
+
+interface State {
+	toasts: ToasterToast[]
+}
+
+const toastTimeouts = new Map<string, ReturnType<typeof setTimeout>>()
+
+const addToRemoveQueue = (toastId: string) => {
+	if (toastTimeouts.has(toastId)) {
+		return
+	}
+
+	const timeout = setTimeout(() => {
+		toastTimeouts.delete(toastId)
+		dispatch({
+			type: "REMOVE_TOAST",
+			toastId: toastId,
+		})
+	}, TOAST_REMOVE_DELAY)
+
+	toastTimeouts.set(toastId, timeout)
+}
+
+export const reducer = (state: State, action: Action): State => {
+	switch (action.type) {
+		case "ADD_TOAST":
+			return {
+				...state,
+				toasts: [action.toast, ...state.toasts].slice(0, TOAST_LIMIT),
+			}
+
+		case "UPDATE_TOAST":
+			return {
+				...state,
+				toasts: state.toasts.map((t) =>
+					t.id === action.toast.id ? { ...t, ...action.toast } : t
+				),
+			}
+
+		case "DISMISS_TOAST": {
```

Comment:
Just for my understanding, will a toast be dismissed only when this code is executed or is there any automatic dismissal timeout as well?

Reply:
For a toast-user POV, you don't have to do anything. You just have to mention the title, description and action in the `toast` function returned by the useToast hook.

Internally, yet, the toast gets dismissed when this code is executed. It gets executed automatically after a timeout - which is cancelled if the user interacts with the toast, or when the user clicks on the cross button on the top right of the toast. Or when the user swipes right on the toast.

You can see these interactions in the GIF that I shared.